### PR TITLE
Retrieve and Parse XML for hardcoded transcript text and timestamp.

### PR DESCRIPTION
**Pull request by amuamushu**
Retrieves and parses the XML with the transcript upon fetching /transcript. 

Transcript (Start time, duration, and text) as printed in the console: https://screenshot.googleplex.com/SNdLcSTDsAN.png

### CODE REVIEW

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * Provides the transcript for the lecture.
+ */
+@WebServlet("/transcript")
+public class TranscriptServlet extends HttpServlet {
+  private static final String TRANSCRIPT_XML_URL_TEMPLATE =
+      "http://video.google.com/timedtext?lang=en&v=";
+  private static final String START_ATTRIBUTE = "start";
+  private static final String DURATION_ATTRIBUTE = "dur";
+  private static final String TEXT_TAG = "text";
+  private static final String TEST_VIDEO_ID = "3ymwOvzhwHs";
```
Comment (jaysaleh) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:45]: I would add a TODO here to get the actual ID

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+package com.googleinterns.zoomtube.servlets;
+
+import java.io.IOException;
+import java.net.URL;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.xml.namespace.QName;
```
Comment (n-wach) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:24]: nit: some unused imports in here

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import javax.xml.parsers.ParserConfigurationException;
+import org.w3c.dom.Attr;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+/**
+ * Provides the transcript for the lecture.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:35]:
```suggestion
 * Provides the transcript for a given lecture.
```

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ */
+@WebServlet("/transcript")
+public class TranscriptServlet extends HttpServlet {
+  private static final String TRANSCRIPT_XML_URL_TEMPLATE =
+      "http://video.google.com/timedtext?lang=en&v=";
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:40]: What is this? Is it 1 video? Do you plan on making this a variable? If so, you should you add a TODO.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  private static final String START_ATTRIBUTE = "start";
+  private static final String DURATION_ATTRIBUTE = "dur";
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:42]: You should not use shortcuts unless very well known. I would say 'dur' is not very common.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+  @Override
+  public void init() throws ServletException {
+    // TODO: Implement Transcript.
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    try {
+      // Later, the video ID will be passed in from another servlet.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:55]: This should be made into a TODO.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      String transcriptXMLUrl = TRANSCRIPT_XML_URL_TEMPLATE + TEST_VIDEO_ID;
+      DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+      DocumentBuilder db = dbf.newDocumentBuilder();
+      Document doc = db.parse(new URL(transcriptXMLUrl).openStream());
+      doc.getDocumentElement().normalize();
+
+      NodeList nodeList = doc.getElementsByTagName(TEXT_TAG);
+      // A for loop is used because NodeList is not an Iterable.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:63]: nit: I do not think this comment is necessary.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java (hunk 5 above)
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:42]: dur is the attribute of an XML tag that was fetched from the API so I didn't decide on what to name it.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java (hunk 4 above)
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:40]: (Resolved in chat) The string is a template for fetching XML. I have to concat the video id to the end of it to fetch whichever video I want.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java (hunk 6 above)
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java]: Done!

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java (hunk 7 above)
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java]: Removed
Done!

**Comment by billie-cat** [comment]
@googlebot I fixed it

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+  @Override
+  public void init() throws ServletException {
+    // TODO: Implement Transcript.
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    try {
+      // TODO: Pass the video ID from another servlet.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:55]: Does all this code need to be in the try statement? It seems a bit large of a try statement?

Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java]: I moved the lines of code for the node outside of the try statement.

Done!

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  private static final String TEST_VIDEO_ID = "3ymwOvzhwHs";
+
+  @Override
+  public void init() throws ServletException {
+    // TODO: Implement Transcript.
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // TODO: Pass the video ID from another servlet.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:54]: I guess this TODO is a duplicate of the TODO on line44. I would remove one or the other.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+    DocumentBuilder db = null;
+    Document doc = null;
+    try {
+      db = dbf.newDocumentBuilder();
+      doc = db.parse(new URL(transcriptXMLUrl).openStream());
+      doc.getDocumentElement().normalize();
+    } catch (ParserConfigurationException | SAXException e) {
+      // TODO: alert the user.
+      System.out.println("XML parsing error");
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:65]: after you catch the error should you return?

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  public void init() throws ServletException {
+    // TODO: Implement Transcript.
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // TODO: Pass the video ID from another servlet.
+    String transcriptXMLUrl = TRANSCRIPT_XML_URL_TEMPLATE + TEST_VIDEO_ID;
+    DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+    DocumentBuilder db = null;
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:57]: Can you try the following:

final DocumentBuilder db;
final Document doc;

Final indicates that it must be initialized.

Comment (f-camp) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:57]: +1 or if anything make them @Nullable to make it compulsory to perform null checks below.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    } catch (ParserConfigurationException | SAXException e) {
+      // TODO: alert the user.
+      System.out.println("XML parsing error");
+    }
+    NodeList nodeList = doc.getElementsByTagName(TEXT_TAG);
+    for (int nodeIndex = 0; nodeIndex < nodeList.getLength(); nodeIndex++) {
+      Node node = nodeList.item(nodeIndex);
+      Element element = (Element) node;
+      String lineStart = element.getAttribute(START_ATTRIBUTE);
+      String lineDuration = element.getAttribute(DURATION_ATTRIBUTE);
```
Comment (f-camp) [inline src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:72]: Should these not be long or of the sort? I reckon these should represent time.

Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:72]: In a later pull request, I updated them to be Date objects. For this pull request, I just wanted to make sure that the XML is being parsed correctly. Sorry for the confusion!

Comment (f-camp) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:72]: Got it!

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java (hunk 9 above)
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:54]: I originally left them both there so I would remember to check both lines, but I just removed them . 

Done!

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java (hunk 10 above)
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:65]: Thanks for catching that! I added a return statement for now.

File: src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java (hunk 11 above)
Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java:57]: It works with final. Thank you!
Done.

**Pull request merged**